### DISCUSSION TITLE
Q&A for MP1

### DISCUSSION THREAD

**Discussion by stan6** [SOEN345-WINTER2024/discussion#5] [Announcements]
1. There is a example test in the Android app that you uploaded on the Github. Is this test a example for task vi?
> task vi. Add one JUnit test for each arithmetic operation (“+”,”-”,”*”,”/”)   (4 points)
``` 
@Test
public void addition_isCorrect() {
         assertEquals(4, 2 + 2);    
}
````
Answer: The test is not an example for task VI. It is a test added by the author of the tutorial. You should create your own test (your test should be different) for task VI.

2. Is the task VII "implement the logic of "." button" the replication of task IV,V?
>iv.Add little code to make the test pass in Step iii. Commit this code to GitHub repository with the commit message “Initial code for passing TDD test”.              (2 points)
v.Refactor the code for adding the functionality to support entering decimal points. In this step, you should remove any code duplication.                                        (2 points)
vii.Implement the logic for the decimal point button “.”             (4 points)

Answer: No. Task VII is not replication of task IV and V. Task IV should have simple code that makes test in Step III passing and this task requires a commit to the repository. Task V clearly states that there should be a refactoring that removes duplication. Task VII requires implementing the complete logic of "." button.   

3. Which lib should we use to testing the onClick() method, I try to google it, but when i using something like ActivityTestRule, the findViewById function didn't work.

Answer: You should refactoring each functionality(like + or -) by extracting methods from the onClick method. Then, you could test the extracted methods easily.

4. Step III said "Add a JUnit test for checking numbers that involves decimal points". 

`@Test
public void addition_isCorrect() {
         assertEquals(1.3, 1 + 0.3);    
}`

is this a valid test for step3 ? . Which file should I write test in ? ExampleUnitTest or ExampleInstrumentTest ?

Answer: No. This is not a valid test for step 3 because it doesn't call any method in the code under test. You should write in ExampleUnitTest since you will be writing unit test.

5. What does implementing the complete logic of "." button mean ?

Answer: It means implementing the decimal point arithmetic operation (+, -, /...). In previous step, only test is added. Now, need to add the implementation.

6. I find that in Android developers' document, there is a way to test button and Android View by "ActivityTestRule" and "Espresso". This way use functions "withId" and "perform" to test the view. But if i import "import android.support.test.rule.ActivityTestRule;" to the test.java in com.bradteachescode.basiccalculator(test), it will show cant resolve symbol. But when i import to com.bradteachescode.basiccalculator(androidtest), it can be resolve and i can new a ActivityTestRule object.

i mean, must i write my junit test in the original file? i want to write my test in com.bradteachescode.basiccalculator(androidtest). i dont know why the original file cant resolve the ActivityTestRule import.

Answer: You need to write unit tests in ExampleUnitTest. the assignment doesn't ask you to test buttons or view. I am not sure how ActivityTestRule comes into play. You extract the methods for the arithmetic operations and test the methods. Maybe you are thinking about this in a too complicated way?

**Comment by ChemsCode** [comment]
Hi, there are some issues with the initial starting code not running. The dependencies used with the provided version of Gradle are deprecated and not compatible with newer versions of Gradle 8.0+. Some dependencies are also not compatible with any versions of Java 11 or above. Are we expected to downgrade our version of Java/Gradle for this assignment?

Thank you!

**Reply by stan6** [reply]
Hi, @ChemsCode. Thank you for the question. I have checked and it seems that starter code is up to date with the [original repository](https://github.com/BradPerkins99/BasicCalculator). Yes, the same as the project, you will need to download and use the correct version specified in the provided Gradle in your computer. Please note that you can keep multiple versions of Java and Gradle. For Java version, you can search online for how to use multiple versions of Java in your OS (For example, [Windows](https://www.happycoders.eu/java/how-to-switch-multiple-java-versions-windows/)). The same with Gradle, you can maintain multiple versions of Gradle (`https://stackoverflow.com/questions/47884051/install-specific-version-of-gradle`).

**Reply by ChemsCode** [reply]
Alright, thank you!

**Reply by stan6** [reply]
Dear @ChemsCode I have checked the repository again. As recommended, you should use the wrapper for Gradle for compilation and running tests. For any Android project, you need to change to the directory where the gradlew file is and run the command.
In the repository, you can find the gradlew file and make it executable and just run `./gradlew tasks` to see all tasks of a gradle project (e.g., build, run tests). For windows, you will use `gradlew.bat tasks ` or you can refer to https://stackoverflow.com/questions/50500325/how-to-execute-gradle-wrapper-gradlew-command-in-android-studio-in-windows on how to execute the gradlew wrapper if you use Android studio in windows.

To check if you have configured things correctly if you use Android studio, make sure that instead of selecting "Import project (Eclipse ADT, Gradle, etc.)," select "Open an existing Android Studio project" then navigate to the topmost build.gradle file of your project.

**Reply by ChemsCode** [reply]
I'll make sure I do, thank you for the feedback👍

**Comment by stan6** [comment]
You are given the starter code which is exactly the same as tutorials.

For the question about whether to extract method, the hint already said you need to extract. I cannot give more hints than that.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by KA-devl** [comment]
Hi, after playing with various JDK versions and gradle versions, i was finally able to run the project. However, i'm still unable to find a fix for the "AppCompatActivity" being unrecognized. Is there anything specific to be able to fix this issue? 
I have tried investigating the issue and found some insights (https://stackoverflow.com/questions/29199891/cannot-resolve-symbol-appcompatactivity), but the problem is still persistent.
![image](https://github.com/SOEN345-WINTER2024/discussion/assets/83729449/0add8ee0-484c-40e2-97ec-63984e34392c)

Update note : The project seems to be working fine, even with the unrecognized class.

**Reply by jcjc1233** [reply]
For this issue go to Project Structure and change the following. Android Gradle Plug-in Version to 4.2.2 and change gradle version to 6.7.1. It will immediately resolve this issue. Hope it helps

**Reply by walidoow** [reply]
Tried this solution and it worked for me.
If like me the user interface does not let you change the Android Gradle Plug-in, try changing it directly in the `BasicCalculator/build.gradle` file.

![image](https://github.com/SOEN345-WINTER2024/discussion/assets/31325808/f349dae2-d2d0-4b95-aa9c-01d4e091e3f5)

**Reply by jcjc1233** [reply]
Yess you are correct I also had to do that if you clone from android studio directly. You also need to change the gradle in gradle-wrapper-properties to 6.7.1 as well. It's in the dependencies

**Comment by stan6** [comment]
Please take note that appcompat is related to the compileSdkVersion and minSdkVersion. Make sure that your selected SDK version is compatible with the supported version by the app.

Regards.
Shin Hwei
________________________________

**Comment by stan6** [comment]
Due to the many questions, I have added a MyShortTutorialBuildAndRunAndroidApp.pdf under the assignments folder of Moodle. Please refer to the steps. I have tested the steps in my own computer and it works correctly.

**Reply by KhalilGarali** [reply]
Hi, I cannot find the pdf you uploaded, unfortunately. Is it possible that it is not accessible by all sections?

**Comment by ramy-att** [comment]
Hi Professor @stan6, are we expected to refactor the entire MainActivity to essentially move the logic into another class? And test this logic by testing each method (add, subtract, multiply, ...)?

**Comment by stan6** [comment]
Dear Ramy,

As we already provided hint about extracting method, the hint is quite clear. Regarding what to be extracted and the name of the method after you have extracted, it is up to you to design as in test driven development.

Regards,
Shin Hwei

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by stan6** [comment]
Hi, the tutorial is only uploading for those who are enrolled in Section UU under the assignment folder in Moodle.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by KhalilGarali** [reply]
Can you share it with section SS as well?

**Comment by stan6** [comment]
Sorry. I am not teaching the class.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by amirhossein942** [comment]
Hello 
I have some issues after cloning going through the instructions PDF file. At this point I have android studio arctic fox version, sdk version 27, java 8, the virtual mobile is setup, and I put the gradlew commands. However when I want to run I get this error: 
Unable to start the daemon process.

The project uses Gradle 4.4 which is incompatible with Java 11 or newer.

Possible solution:
 - Upgrade Gradle wrapper to 4.8 version and re-import the project
 
 and additionally my gradle project settings does not even have this project for me to change the jdk version there.
 
![photo_2024-02-09_21-46-58](https://github.com/SOEN345-WINTER2024/discussion/assets/83388763/54e1e7aa-cc6d-4d4c-9d82-9e7a2b78a82b)

Not sure how I can change the jdk to 8 for this. 
NOTE: both my JAVA_HOME environment variable and Android Studio's jdk path are set to java 8

**Reply by jcjc1233** [reply]
If you have discord and you follow the SOEN 345 channel, find my username jc1233 on discord. Or just add me there. I can definitely help you fix this

**Reply by amirhossein942** [reply]
Sure thing, I sent you a message on discord

**Comment by thaneekan** [comment]
Hi, I'm not sure how to exactly resolve this. The app did initially build but now, the build has failed because there are duplicate classes between hamcrest and junit. I don't remember enabling hamcrest and not sure how to get rid of it.
![Screenshot (338)](https://github.com/SOEN345-WINTER2024/discussion/assets/92413776/5a6e467d-a0c9-44b1-a5e4-1906d568e20b)

**Reply by shtest1** [reply]
you need to do gradlew clean and rebuild. You also check your build.gradle file and your build path to see if it is included twice.

**Reply by thaneekan** [reply]
hi, even after running the command, the same error shows. I don't see anything included twice in the build.gradle file for the app.
![Screenshot (340)](https://github.com/SOEN345-WINTER2024/discussion/assets/92413776/45a4f75f-fb3e-407c-bccb-8e398957fc3b)

**Comment by shtest1** [comment]
I suggest that you do a search in your entire project for "hamcrest" and remove all the import/dependency in java file and gradle before cleaning and rebuilding. 

If this doesn't work, you can clone the repository from GitHub into another folder and rebuild.

**Reply by thaneekan** [reply]
after searching for hamcrest, this is what shows up. it doesn't really show if it is in within the project or not. I will have to download from the repo with my last commits since even cloning is an issue right now.
![Screenshot (341)](https://github.com/SOEN345-WINTER2024/discussion/assets/92413776/9822d100-8191-4d1b-ab88-4df98f9ff4b5)

**Comment by daria-koroleva** [comment]
For question ix: "Add one JUnit test for checking invalid input and commit this test." Should i add  the code to make the test pass as TDD indicates? or left the test to fail without adding new code? Because pass that question, there are no mention of additional commits?

**Reply by jcjc1233** [reply]
When you add the test it has to fail because it's not fully implemented yet. So you commit it first. Then in the next step you implement the method in the main so you can actually test it.

**Reply by daria-koroleva** [reply]
Thank you, I am asking because after the step ix, there is no a new step indicating  to add the code, like in previous ones. After step ix is just README file update. That's why I want to confirm.

**Reply by jcjc1233** [reply]
You have to commit every single change you make to the code. Commit in every step of the way

**Reply by daria-koroleva** [reply]
assignment said to follow step-by step instructions for each commit, not every step requires commit. So, I only committed as requested.

**Comment by stan6** [comment]
Q: Hi, for mp1 is it okay to clone the original repo of the creator instead of one assigned to us via classroom ? im having issues cloning the classroom repo via android studio

Q: Hi, you can download the GitHub classroom repo as zip from the GitHub if you have issue cloning. We do not recommend cloning the original repo as it may not be in sync.
